src/battle_ai_main.c

```diff
@@ -2645,6 +2645,13 @@ static s32 AI_CheckBadMove(u32 battlerAtk, u32 battlerDef, u32 move, s32 score)
             if (IsDynamaxed(battlerDef))
                 ADJUST_SCORE(-10);
             break;
+        case EFFECT_UPPER_HAND:
+            if (predictedMove != MOVE_NONE)
```

Comment:
You need a case for never using Upper Hand if there is no predicted move. As in you can simplify this to fail if `predictedMove == MOVE_NONE || IS_MOVE_STATUS(predictedMove) || ...`